packages/backend/src/connector/core/middleware/balance.ts

```diff
@@ -44,30 +45,45 @@ export function createBalanceMiddleware(): ILPMiddleware {
     }
 
     // Update balances on prepare
-    const trxOrError = await services.accounting.createTransfer({
-      sourceAccount: accounts.incoming,
-      destinationAccount: accounts.outgoing,
-      sourceAmount,
-      destinationAmount: destinationAmountOrError,
-      timeout: BigInt(5e9) // 5 seconds
-    })
+    const createPendingTransfer = async (): Promise<Transaction> => {
+      const trxOrError = await services.accounting.createTransfer({
+        sourceAccount: accounts.incoming,
+        destinationAccount: accounts.outgoing,
+        sourceAmount,
+        destinationAmount: destinationAmountOrError,
+        timeout: BigInt(5e9) // 5 seconds
+      })
 
-    if (isTransferError(trxOrError)) {
-      switch (trxOrError) {
-        case TransferError.InsufficientBalance:
-        case TransferError.InsufficientLiquidity:
-          throw new InsufficientLiquidityError(trxOrError)
-        default:
-          // TODO: map transfer errors to ILP errors
-          ctxThrow(500, destinationAmountOrError.toString())
+      if (isTransferError(trxOrError)) {
+        switch (trxOrError) {
+          case TransferError.InsufficientBalance:
+          case TransferError.InsufficientLiquidity:
+            throw new InsufficientLiquidityError(trxOrError)
+          default:
+            // TODO: map transfer errors to ILP errors
+            ctxThrow(500, destinationAmountOrError.toString())
+        }
+      } else {
+        return trxOrError
       }
-    } else {
+    }
+
+    if (state.streamDestination) {
       await next()
+    }
+
+    if (!state.streamDestination || response.fulfill) {
+      // TODO: make this single-phase if streamDestination === true
+      const trx = await createPendingTransfer()
+
+      if (!state.streamDestination) {
+        await next()
+      }
```

Comment:
>it seems await next() will be called regardless if there is or isn't a streamDestination

Correct.
The key change is that `createPendingTransfer` now isn't called if `state.streamDestination && !response.fulfill`.

>Do createPendingTransfer() side-effects cause changes downstream in next() ?

It is important to call `createPendingTransfer` before `next` when this isn't the stream destination, because we don't want to forward the ILP packet to subsequent connectors (in subsequent middlewares) if a local balance is insufficient.